packages/core/src/utils/storage/s3-storage.ts

```diff
@@ -2,17 +2,38 @@ import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
 
 import type { UploadFile } from './types.js';
 
-export const buildS3Storage = (
-  endpoint: string,
-  bucket: string,
-  accessKeyId: string,
-  secretAccessKey: string
-) => {
+const getRegionFromEndpoint = (endpoint?: string) => {
+  if (!endpoint) {
+    return;
+  }
+
+  return /s3\.([^.]*)\.amazonaws/.exec(endpoint)?.[1];
+};
+
+type BuildS3StorageParameters = {
+  bucket: string;
+  accessKeyId: string;
+  secretAccessKey: string;
+  region?: string;
+  endpoint?: string;
+};
+
+export const buildS3Storage = ({
+  bucket,
+  accessKeyId,
+  secretAccessKey,
+  region,
+  endpoint,
+}: BuildS3StorageParameters) => {
+  if (!region && !endpoint) {
+    throw new Error('Either region or endpoint must be provided');
+  }
+
   // Endpoint example: s3.us-west-2.amazonaws.com
-  const region = /s3\.([^.]*)\.amazonaws/.exec(endpoint)?.[1] ?? 'us-east-1';
+  const finalRegion = region ?? getRegionFromEndpoint(endpoint) ?? 'us-east-1';
```

Comment:
If the specified region does not match the one from endpoint, should we warn the user or even throw an error?

Reply:
But I am afraid that would be a unnecessary breaking change, some providers DO NOT need `region`, the SDK need `region`, some documentation also says that the if you don't care about region, just set it to `us-east-1`.

Reply:
ok